chpl/chpl-service/src/main/java/gov/healthit/chpl/activity/history/ListingActivityHistoryHelper.java

```diff
@@ -0,0 +1,111 @@
+        if (StringUtils.isBlank(listing.getSvapNoticeUrl())) {
+            LOGGER.info("No SVAP Notice URL for listing ID " + listing.getId());
+            return null;
+        }
+
+        LOGGER.info("Getting last update date for SVAP Notice URL for listing ID " + listing.getId() + ".");
+        List<ActivityDTO> listingActivities = activityDao.findByObjectId(listing.getId(), ActivityConcept.CERTIFIED_PRODUCT, EPOCH, new Date());
+        if (listingActivities == null || listingActivities.size() == 0) {
+            LOGGER.warn("No listing activities were found for listing ID " + listing.getId() + ". Is the ID valid?");
+            return null;
+        }
+        LOGGER.info("There are " + listingActivities.size() + " activities for listing ID " + listing.getId());
+        activityUtil.sortNewestActivityFirst(listingActivities);
+
+        ActivityDTO svapNoticeLastUpdateActivity = null;
+        Iterator<ActivityDTO> listingActivityIter = listingActivities.iterator();
+        while (svapNoticeLastUpdateActivity == null && listingActivityIter.hasNext()) {
+            ActivityDTO currActivity = listingActivityIter.next();
+            CertifiedProductSearchDetails orig = null, updated = null;
+            if (currActivity.getOriginalData() != null) {
+                orig = activityUtil.getListing(currActivity.getOriginalData());
+            }
+            if (currActivity.getNewData() != null) {
+                updated = activityUtil.getListing(currActivity.getNewData());
+            }
+
+            if (wasSvapNoticeUrlSetToCurrent(orig, updated, listing.getSvapNoticeUrl())) {
+                svapNoticeLastUpdateActivity = currActivity;
+                LOGGER.info("Listing " + listing.getId() + " SVAP Notice URL was set to " + listing.getSvapNoticeUrl() + " on " + currActivity.getActivityDate() + ".");
+            }
+        }
+
+        if (svapNoticeLastUpdateActivity == null) {
+            LOGGER.warn("Unable to determine when listing " + listing.getId() + " set SVAP Notice URL to " + listing.getSvapNoticeUrl());
+        }
+        return svapNoticeLastUpdateActivity;
+    }
+
+    private boolean wasSvapNoticeUrlSetToCurrent(CertifiedProductSearchDetails orig, CertifiedProductSearchDetails updated,
+            String currentSvapNoticeUrl) {
+        if (orig != null && updated != null && !StringUtils.isBlank(updated.getSvapNoticeUrl())
+                && updated.getSvapNoticeUrl().equals(currentSvapNoticeUrl)
+                && (orig.getSvapNoticeUrl() == null || !currentSvapNoticeUrl.equals(orig.getSvapNoticeUrl()))) {
+            return true;
+        } else if (orig == null && updated != null && !StringUtils.isBlank(updated.getSvapNoticeUrl())
+                && updated.getSvapNoticeUrl().equals(currentSvapNoticeUrl)) {
+            return true;
+        }
+        return false;
+    }
+
+    @Transactional
+    public CertifiedProductSearchDetails getListingOnDate(Long listingId, Date date) {
```

Comment:
Would it make sense to move the SVAP helper functions into a different class? This method looks like a "ListingActivityHistoryHelper" method, but the other ones look very specific for a particular type of history